site/src/pages/WorkspacePage/WorkspaceNotifications.tsx

```diff
@@ -0,0 +1,343 @@
+      detail: workspace.deleting_at ? (
+        <>
+          This workspace has not been used for{" "}
+          {formatDistanceToNow(Date.parse(workspace.last_used_at))} and was
+          marked dormant on {formatDate(workspace.dormant_at, false)}. It is
+          scheduled to be deleted on {formatDate(workspace.deleting_at, true)}.
+          To keep it you must activate the workspace.
+        </>
+      ) : (
+        <>
+          This workspace has not been used for{" "}
+          {formatDistanceToNow(Date.parse(workspace.last_used_at))} and was
+          marked dormant on {formatDate(workspace.dormant_at, false)}. It is not
+          scheduled for auto-deletion but will become a candidate if
+          auto-deletion is enabled on this template. To keep it you must
+          activate the workspace.
+        </>
+      ),
+    });
+  }
+
+  // Pending in Queue
+  const [showAlertPendingInQueue, setShowAlertPendingInQueue] = useState(false);
+  // 2023-11-15 - MES - This effect will be called every single render because
+  // "now" will always change and invalidate the dependency array. Need to
+  // figure out if this effect really should run every render (possibly meaning
+  // no dependency array at all), or how to get the array stabilized (ideal)
+  const now = dayjs();
+  useEffect(() => {
+    if (
+      workspace.latest_build.status !== "pending" ||
+      workspace.latest_build.job.queue_size === 0
+    ) {
+      if (!showAlertPendingInQueue) {
+        return;
+      }
+
+      const hideTimer = setTimeout(() => {
+        setShowAlertPendingInQueue(false);
+      }, 250);
+      return () => {
+        clearTimeout(hideTimer);
+      };
+    }
+
+    const t = Math.max(
+      0,
+      5000 - dayjs().diff(dayjs(workspace.latest_build.created_at)),
+    );
+    const showTimer = setTimeout(() => {
+      setShowAlertPendingInQueue(true);
+    }, t);
+
+    return () => {
+      clearTimeout(showTimer);
+    };
+  }, [workspace, now, showAlertPendingInQueue]);
+
+  if (showAlertPendingInQueue) {
+    notifications.push({
+      title: "Workspace build is pending",
+      severity: "info",
+      detail: (
+        <>
+          <div css={{ marginBottom: 12 }}>
+            This workspace build job is waiting for a provisioner to become
+            available. If you have been waiting for an extended period of time,
+            please contact your administrator for assistance.
+          </div>
+          <div>
+            Position in queue:{" "}
+            <strong>{workspace.latest_build.job.queue_position}</strong>
+          </div>
+        </>
+      ),
+    });
+  }
+
+  // Deprecated
+  if (template.deprecated) {
+    notifications.push({
+      title: "Workspace using deprecated template",
+      severity: "warning",
+      detail: template.deprecation_message,
+    });
+  }
+
+  const infoNotifications = notifications.filter((n) => n.severity === "info");
+  const warningNotifications = notifications.filter(
```

Comment:
I'm wondering if it's practical/a good idea to try splitting up the notifications logic into a helper function. Something like this:
```ts
function getNotifications (...inputs) {

  /** Main logic goes here */

  return {
    info: [...],
    warnings: [...],
  };
}
```

My main worry is that a lot of the branches are kind of interspersed throughout the component, with a random `useEffect` call wedged in there

Reply:
I'm not sure if I got it 🤔

Reply:
I know the code has changed a little bit since the comment, but I'm going to go with the code on this PR as an example. The way I see the `WorkspaceNotifications` component, it kind of has different sections:

```diff
export const WorkspaceNotifications = (props) => {
  <-- Props destructuring -->
  <-- Set up empty notifications array -->
  <-- Calculate values from workspace query -->
  <-- Mutate notifications array if outdated -->
  <-- Mutate notifications array if unhealthy -->
  <-- Set up state for checking whether actions are enabled -->
  <-- Mutate notifications array if actions are enabled -->
  <-- Set up state/effects for alerts -->
  <-- Mutate notifications array if alerts are pending -->
  <-- Mutate notifications array if template is deprecated -->
  <-- Mutate notifications array if template is deprecated -->
  <-- Split the notifications array by warning/info type -->
  <-- Return JSX output -->
};
```
Most of the sections are dealing with calculating the notifications array
```diff
export const WorkspaceNotifications = (props) => {
   <-- Props destructuring -->
+  <-- Set up empty notifications array -->
   <-- Calculate values from workspace query -->
+  <-- Mutate notifications array if outdated -->
+  <-- Mutate notifications array if unhealthy -->
   <-- Set up state for checking whether actions are enabled -->
+  <-- Mutate notifications array if actions are enabled -->
   <-- Set up state/effects for alerts -->
+  <-- Mutate notifications array if alerts are pending -->
+  <-- Mutate notifications array if template is deprecated -->
+  <-- Mutate notifications array if template is deprecated -->
+  <-- Split the notifications array by warning/info type -->
   <-- Return JSX output -->
};
```
But the notifications array should always be the same if we have the same inputs, so seemingly, there should be a way to extract it into a reusable function. If we had the function, we could reorganize things like this:
```diff
export const WorkspaceNotifications = (props) => {
   <-- Props destructuring -->
   <-- Calculate values from workspace query -->
   <-- Set up state for checking whether actions are enabled -->
   <-- Set up state/effects for alerts -->
+  <-- Call function to get notifications -->
   <-- Return JSX output -->
};
```
The function might still mutate an array under the hood to fill it with data, but as long as we don't have any side effects, it will be fine, and it might be easier to add more cases for adding extra notifications down the line

Reply:
But for some of them, we need to fetch data... so how to deal with that without any effect?

Reply:
We would still have the effects and fetching – that's why I didn't remove the non-notification steps. It's just that their values would get passed into the function on each render

I did just realize I forgot some sections, though. It should be:
```diff
export const WorkspaceNotifications = (props) => {
   <-- Props destructuring -->
+  <-- Set up workspaces autostart query state -->
   <-- Calculate values from workspace query -->
   <-- Set up state for checking whether actions are enabled -->
   <-- Set up state/effects for alerts -->
+  <-- Call function to get notifications -->
   <-- Return JSX output -->
};
```